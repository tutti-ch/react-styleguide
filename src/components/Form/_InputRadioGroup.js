import React, { PureComponent, cloneElement, Children } from "react";
import PropTypes from "prop-types";
import WithWrapper from "./_WithWrapper";
import classes from "./Form.scss";

export class InputRadioGroup extends PureComponent {
  static propTypes = {
    children: PropTypes.node,

    /**
     * The on change callback handler.
     */
    onChange: PropTypes.func,

    /**
     * A name to pass to children. If you provide a name
     * attribute on the child, it will overwrite.
     */
    name: PropTypes.string,

    /**
     * Label to show above the options
     */
    label: PropTypes.string
  };

  constructor(props) {
    super(props);

    const state = {
      selected: null,
      value: undefined
    };

    Children.map(this.props.children, (child, index) => {
      if (child.props.checked) {
        state.selected = index;
      }
    });

    this.state = state;
  }

  /**
   * Handle the on change.
   *
   * @param index
   * @return {function()}
   */
  handleOnChange(index) {
    return (value, props) => {
      const { onChange } = this.props;

      if (typeof onChange === "function") {
        onChange(value, props);
      }

      this.setState({ selected: index, value });
    };
  }

  render() {
    const { label } = this.props;
    return (
      <div>
        {label && <span className={classes.radioLabel}>{label}</span>}
        {Children.map(this.props.children, (child, index) =>
          cloneElement(child, {
            checked: this.state.selected === index,
            onChange: this.handleOnChange(index),
            name: this.props.name || child.props.name // Radio group has a precedence
          })
        )}
      </div>
    );
  }
}

export default WithWrapper(InputRadioGroup);
