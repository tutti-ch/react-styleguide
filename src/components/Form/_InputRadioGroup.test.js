/* global describe, test, expect, jest */
import React from "react";
import { InputRadioGroup } from "./_InputRadioGroup";
import InputRadio from "./_InputRadio";
import { mount } from "enzyme";

describe("(Component) InputRadioGroup", () => {
  test("should handle the change of the children states", () => {
    const onChange = jest.fn();

    const comp = mount(
      <InputRadioGroup onChange={onChange}>
        <InputRadio label="React.js" name="fmw" />
        <InputRadio label="Vue.js" name="fmw" checked />
        <InputRadio label="Angular.js" value="ang" name="fmw" />
      </InputRadioGroup>
    );

    expect(comp.state("selected")).toBe(1); // The index of the selected radio

    comp.instance().handleOnChange(2)("ang", { name: "fmw" });
    expect(onChange).toHaveBeenCalledWith("ang", { name: "fmw" });
    expect(comp.state("selected")).toBe(2);

    comp.update();
    expect(comp).toMatchSnapshot();
  });
});
