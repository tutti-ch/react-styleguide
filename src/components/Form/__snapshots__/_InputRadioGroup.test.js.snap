// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) InputRadioGroup should handle the change of the children states 1`] = `
<InputRadioGroup
  onChange={[Function]}
>
  <div>
    <InputRadio
      checked={false}
      key=".0"
      label="React.js"
      name="fmw"
      onChange={[Function]}
    >
      <span
        onClick={[Function]}
      >
        <input
          checked={false}
          name="fmw"
          onChange={[Function]}
          type="radio"
        />
        <span>
          React.js
        </span>
      </span>
    </InputRadio>
    <InputRadio
      checked={false}
      key=".1"
      label="Vue.js"
      name="fmw"
      onChange={[Function]}
    >
      <span
        onClick={[Function]}
      >
        <input
          checked={false}
          name="fmw"
          onChange={[Function]}
          type="radio"
        />
        <span>
          Vue.js
        </span>
      </span>
    </InputRadio>
    <InputRadio
      checked={true}
      key=".2"
      label="Angular.js"
      name="fmw"
      onChange={[Function]}
      value="ang"
    >
      <span
        onClick={[Function]}
      >
        <input
          checked={true}
          name="fmw"
          onChange={[Function]}
          type="radio"
        />
        <span>
          Angular.js
        </span>
      </span>
    </InputRadio>
  </div>
</InputRadioGroup>
`;
